nodeJS/apis/RESTful_APIs.md

```diff
@@ -0,0 +1,127 @@
+- `501 Not Implemented`: The server lacks the capabilities to fulfill the request or is unable to recognize the request method.
+
+### URLs
+
+By now, I'm sure we know what URLs are, but as a quick refresher, let's check one out. It consists of a simple structure of a few components. Take the following as an example.
+
+```text
+http://sub.domain.com:1234/path/to/resource?query=something&param=something#anchor
+```
+
+- Here, `http` is the protocol. Some others that you might notice include https, ftp, and so on.
+- The hostname is `sub.domain.com`
+- The subdomain here is `sub`
+- The domain name is `domain.com`
+- `1234` indicates the port. The browser usually hides this, since it defaults to `80`, which is the default port for web servers.
+- `/path/to/resource` denotes the resource path which refers to a resource location within the server.
+- There are some query parameters here: `?query=something&param=something`. Query string parameters start with a question mark (`?`), following a pattern of `key=value` pairs, which are separated by ampersands (`&`). There can be any number of parameters, and they are generally used by the server to identify the right resource.
+- The `#anchor` is called a fragment or a named anchor. It is used to refer to an internal section within a web document.
+
+### RESTful APIs
+
+Simply put, an API is an interface. When an application needs to interact with another, it sends a request to the respective API. As you've learned in previous lessons, in the context of the web, any server that is created to serve data for external use is called an API. While you can structure your API in multiple ways, a popular and conventional method to do so is to follow REST (**Re**presentational **S**tate **T**ransfer). [The exact definition of REST](https://en.wikipedia.org/wiki/REST#Principle) might be a little complicated, but for us, it states that there is a set of standards to be followed to make our API RESTful (adhere to the constraints set by REST).
+
+Earlier, we mentioned separating the client and the server, which fulfills the first constraint of REST: the frontend and backend are well-defined. Further constraints like statelessness and caching are covered by and ensured later with ExpressJS. Our key concern at this point is the organization of endpoint URLs with respect to our resources.
+
+As mentioned, usually your backend application will need to send data to your frontend. The most popular way to do so by far is with JSON, primarily due to the ease of parsing it with JavaScript. So all we need to do is to replace our HTML and serve JSON instead. All that you have to do, thus, is to pass your information to [`res.json()`](https://expressjs.com/en/4x/api.html#res.json) instead of [`res.send()`](https://expressjs.com/en/4x/api.html#res.send) or [`res.render()`](https://expressjs.com/en/4x/api.html#res.render).
```

Comment:
I'm still debating this but what do we think about having this section start the lesson instead of it being in the end?

Reply:
sounds good to me